File: examples/bi-teleoperation/flexiv_rdk-main/example_py/basics7_auto_recovery.py
#!/usr/bin/env python

"""basics7_auto_recovery.py

This tutorial runs an automatic recovery process if the robot's safety system is in recovery
state. See flexiv::Robot::isRecoveryState() and RDK manual for more details.
"""

__copyright__ = "Copyright (C) 2016-2021 Flexiv Ltd. All Rights Reserved."
__author__ = "Flexiv"

import time
import argparse

# Import Flexiv RDK Python library
# fmt: off
import sys
sys.path.insert(0, "../lib_py")
import flexivrdk
# fmt: on

# Global flag: whether the gripper control tasks are finished
g_is_done = False


def print_description():
    """
    Print tutorial description.

    """
    print("This tutorial runs an automatic recovery process if the robot's safety system is in "
          "recovery state. See flexiv::Robot::isRecoveryState() and RDK manual for more details.")
    print()


def main():
    # Program Setup
    # ==============================================================================================
    # Parse arguments
    argparser = argparse.ArgumentParser()
    argparser.add_argument('robot_ip', help='IP address of the robot server')
    argparser.add_argument('local_ip', help='IP address of this PC')
    args = argparser.parse_args()

    # Define alias
    log = flexivrdk.Log()

    # Print description
    log.info("Tutorial description:")
    print_description()

    try:
        # RDK Initialization
        # ==========================================================================================
        # Instantiate robot interface
        robot = flexivrdk.Robot(args.robot_ip, args.local_ip)

        # Enable the robot, make sure the E-stop is released before enabling
        log.info("Enabling robot ...")
        robot.enable()

        # Run Auto-recovery
        # ==========================================================================================
        # If the system is in recovery state, we can't use isOperational to tell if the enabling
        # process is done, so just wait long enough for the process to finish
        time.sleep(8)

        # Start auto recovery if the system is in recovery state, the involved joints will start to
        # move back into allowed position range
        if robot.isRecoveryState():
            robot.startAutoRecovery()
            # Block forever, must reboot the robot and restart user program after recovery is done
            while True:
                time.sleep(1)

        # Otherwise the system is normal, do nothing
        else:
            log.info(
                "Robot system is not in recovery state, nothing to be done, exiting ...")

    except Exception as e:
        # Print exception error message
        log.error(str(e))


if __name__ == "__main__":
    main()

File: examples/bi-teleoperation/flexiv_rdk-main/example/basics7_auto_recovery.cpp
/**
 * @example basics7_auto_recovery.cpp
 * This tutorial runs an automatic recovery process if the robot's safety system is in recovery
 * state. See flexiv::Robot::isRecoveryState() and RDK manual for more details.
 * @copyright Copyright (C) 2016-2021 Flexiv Ltd. All Rights Reserved.
 * @author Flexiv
 */

#include <flexiv/Robot.hpp>
#include <flexiv/Exception.hpp>
#include <flexiv/Log.hpp>
#include <flexiv/Utility.hpp>

#include <iostream>
#include <string>
#include <thread>

/** @brief Print tutorial description */
void printDescription()
{
    std::cout
        << "This tutorial runs an automatic recovery process if the robot's safety system is in "
           "recovery state. See flexiv::Robot::isRecoveryState() and RDK manual for more details."
        << std::endl
        << std::endl;
}

/** @brief Print program usage help */
void printHelp()
{
    // clang-format off
    std::cout << "Required arguments: [robot IP] [local IP]" << std::endl;
    std::cout << "    robot IP: address of the robot server" << std::endl;
    std::cout << "    local IP: address of this PC" << std::endl;
    std::cout << "Optional arguments: None" << std::endl;
    std::cout << std::endl;
    // clang-format on
}

int main(int argc, char* argv[])
{
    // Program Setup
    // =============================================================================================
    // Logger for printing message with timestamp and coloring
    flexiv::Log log;

    // Parse parameters
    if (argc < 3 || flexiv::utility::programArgsExistAny(argc, argv, {"-h", "--help"})) {
        printHelp();
        return 1;
    }
    // IP of the robot server
    std::string robotIP = argv[1];
    // IP of the workstation PC running this program
    std::string localIP = argv[2];

    // Print description
    log.info("Tutorial description:");
    printDescription();

    try {
        // RDK Initialization
        // =========================================================================================
        // Instantiate robot interface
        flexiv::Robot robot(robotIP, localIP);

        // Enable the robot, make sure the E-stop is released before enabling
        log.info("Enabling robot ...");
        robot.enable();

        // Run Auto-recovery
        // =========================================================================================
        // If the system is in recovery state, we can't use isOperational to tell if the enabling
        // process is done, so just wait long enough for the process to finish
        std::this_thread::sleep_for(std::chrono::seconds(8));

        // Start auto recovery if the system is in recovery state, the involved joints will start to
        // move back into allowed position range
        if (robot.isRecoveryState()) {
            robot.startAutoRecovery();
            // Block forever, must reboot the robot and restart user program after recovery is done
            while (true) {
                std::this_thread::sleep_for(std::chrono::seconds(1));
            }
        }
        // Otherwise the system is normal, do nothing
        else {
            log.info("Robot system is not in recovery state, nothing to be done, exiting ...");
        }
    } catch (const flexiv::Exception& e) {
        log.error(e.what());
        return 1;
    }

    return 0;
}
